include/tscpp/api/HttpMsgComp.h

```diff
@@ -0,0 +1,424 @@
+  MimeField(const MimeField &) = delete;
+  MimeField &operator=(const MimeField &) = delete;
+
+  MimeField(MimeField &&source);
+
+  // Move assign only allowed if the two instances refer to the same message.
+  //
+  MimeField &operator=(MimeField &&source);
+
+  void destroy();
+
+  // Next field, returns invalid instance if none.
+  //
+  MimeField next() const;
+
+  // Next field with same name, return invalid instance if none.
+  //
+  MimeField nextDup() const;
+
+  std::string_view nameGet() const;
+
+  void nameSet(std::string_view);
+
+  void valuesClear();
+
+  // Returns number of values. Values index are from 0 to valuesCount() - 1.
+  //
+  int valuesCount() const;
+
+  std::string_view valGet(int idx) const;
+
+  void valSet(int idx, std::string_view new_value);
+
+  // Insert a new value at index idx.  If idx == valuesCount(), the new value is appended at the end.  Otherwise, all
+  // values with index >= idx prior to calling this have their index incremented by one.
+  //
+  void valInsert(int idx, std::string_view new_value);
+
+  ~MimeField();
+
+private:
+  MsgBase &_msg;
+  TSMLoc _loc;
+};
+
+class MsgBase : public MsgBuffer
+{
+public:
+  MsgBase(TSMBuffer bHndl, TSMLoc msgLoc) : MsgBuffer(bHndl), _msgLoc(msgLoc) {}
+
+  TSMLoc
+  msgLoc() const
+  {
+    return _msgLoc;
+  }
+
+  bool
+  hasMsg() const
+  {
+    return msgLoc() != TS_NULL_MLOC;
+  }
+
+  //  Returns number of mime header lines in HTTP message.  Can only be called if hasMsg() is true.
+  //
+  int
+  mimeFieldsCount() const
+  {
+    assert(hasMsg());
+
+    return TSMimeHdrFieldsCount(bufHndl(), msgLoc());
+  }
+
+  MimeField
+  newMimeField()
+  {
+    assert(hasMsg());
+
+    TSMLoc loc;
+    if (TSMimeHdrFieldCreate(bufHndl(), msgLoc(), &loc) != TS_SUCCESS) {
+      loc = TS_NULL_MLOC;
+    }
+    return MimeField(*this, loc);
+  }
+
+  ~MsgBase()
+  {
+    if (_msgLoc && bufHndl()) {
+      assert(TSHandleMLocRelease(bufHndl(), TS_NULL_MLOC, _msgLoc) == TS_SUCCESS);
+    }
+  }
+
+protected:
+  MsgBase() : MsgBuffer(nullptr), _msgLoc(TS_NULL_MLOC) {}
+
+  void
+  setMsgLoc(TSMLoc msgLoc)
+  {
+    _msgLoc = msgLoc;
+  }
+
+private:
+  TSMLoc _msgLoc;
+};
+
+template <TSReturnCode (*getInTxn)(TSHttpTxn txnp, TSMBuffer *, TSMLoc *)> class TxnMsg : public MsgBase
+{
+public:
+  TxnMsg() : MsgBase(nullptr, TS_NULL_MLOC) {}
+
+  // Acutal type of txnHndl must be TSHttpTxn.  Using void * for compatability with
+  // atscppapi::Transaction::getAtsHandle().
+  //
+  explicit TxnMsg(void *txnHndl) { _init(txnHndl); }
+
+  // Must not be called on instance that already has a message.
+  //
+  bool
+  init(void *txnHndl)
+  {
+    assert(!hasMsg());
+
+    _init(txnHndl);
+
+    return (hasMsg());
+  }
+
+private:
+  void
+  _init(void *txnHndl)
+  {
+    TSMBuffer msgBuffer;
+    TSMLoc msgLoc;
+
+    if (TS_SUCCESS == getInTxn(static_cast<TSHttpTxn>(txnHndl), &msgBuffer, &msgLoc)) {
+      set(msgBuffer);
+      setMsgLoc(msgLoc);
+    }
+  }
+};
+
+#undef X
```

Comment:
I think this makes the code less clear, since you must mentally decode the macro in order to know what's been defined.

Reply:
Mental exercise delays cognitive decline.

Reply:
The problem with this way of thinking is, where do you draw the line?  Suppose someone says they don't want to figure out what std::unorder_map is going to do with a particular pair of key, value types, they'd rather just write their own untempleted, type-specific hash table?  Macros are much much simpler than templates.

We are all too familiar with the term "cut-n-paste error".  A good reason to use this alternative to cutting and pasting when possible.

The main problem with macros is that they are not scoped.  But that doesn't pertain here, since the macro only exists within a single bottom-level scope.

Reply:
I don't think it's that hard. You're contrasting this case, which is entirely about typing, with a very different case of using different data structures. That's a false equivalence. This is much more akin to a debate about `using`.

Another substantive objection here is that if someone uses one of the defined symbols, how can they find the declaration? That text is nowhere in the code base.

Reply:
You never let me have any fun.